internal/enginenetx/httpsdialer.go

```diff
@@ -0,0 +1,506 @@
+		NextProtos:         []string{"h2", "http/1.1"},
+		RootCAs:            hd.rootCAs,
+		ServerName:         tactic.SNI(),
+	}
+
+	// create handshaker and establish a TLS connection
+	ol = logx.NewOperationLogger(
+		logger,
+		"TLSHandshake with %s SNI=%s ALPN=%v",
+		endpoint,
+		tlsConfig.ServerName,
+		tlsConfig.NextProtos,
+	)
+	thx := tactic.NewTLSHandshaker(netx, logger)
+	tlsConn, err := thx.Handshake(ctx, tcpConn, tlsConfig)
+	ol.Stop(err)
+
+	// handle handshake error
+	if err != nil {
+		tactic.OnTLSHandshakeError(err)
+		tcpConn.Close()
+		return nil, err
+	}
+
+	// verify the certificate chain
+	ol = logx.NewOperationLogger(logger, "TLSVerifyCertificateChain %s", tactic.VerifyHostname())
+	err = httpsDialerVerifyCertificateChain(tactic.VerifyHostname(), tlsConn, hd.rootCAs)
+	ol.Stop(err)
+
+	// handle verification error
+	if err != nil {
+		tactic.OnTLSVerifyError(err)
+		tlsConn.Close()
+		return nil, err
+	}
+
+	// make sure the tactic know it worked
+	tactic.OnSuccess()
+
+	return tlsConn, nil
+}
+
+// httpsDialerWaitReady waits for the given delay to expire or the context to be canceled. If the
+// delay is zero or negative, we immediately return nil. We also return nil when the delay expires. We
+// return the context error if the context expires.
+func httpsDialerTacticWaitReady(ctx context.Context, tactic HTTPSDialerTactic) error {
+	delay := tactic.InitialDelay()
+	if delay <= 0 {
+		return nil
+	}
+
+	timer := time.NewTimer(delay)
+	defer timer.Stop()
+
+	select {
+	case <-timer.C:
+		return nil
+
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// httpsDialerVerifyCertificateChain verifies the certificate chain with the given hostname.
+func httpsDialerVerifyCertificateChain(hostname string, conn model.TLSConn, rootCAs *x509.CertPool) error {
+	// This code comes from the example in the Go source tree that shows
+	// how to override certificate verification and which is advertised
+	// as follows:
+	//
+	//	VerifyConnection can be used to replace and customize connection
+	//	verification. This example shows a VerifyConnection implementation that
+	//	will be approximately equivalent to what crypto/tls does normally to
+	//	verify the peer's certificate.
+	//
+	// See https://github.com/golang/go/blob/go1.21.0/src/crypto/tls/example_test.go#L186
+	//
+	// As of go1.21.0, the code we're replacing has approximately the same
+	// implementation of the verification code we added below.
+	//
+	// See https://github.com/golang/go/blob/go1.21.0/src/crypto/tls/handshake_client.go#L962.
+
+	state := conn.ConnectionState()
+	opts := x509.VerifyOptions{
+		DNSName:       hostname, // note: here we're using the real hostname
+		Intermediates: x509.NewCertPool(),
+		Roots:         rootCAs,
+	}
+	for _, cert := range state.PeerCertificates[1:] {
+		opts.Intermediates.AddCert(cert)
+	}
+
+	if _, err := state.PeerCertificates[0].Verify(opts); err != nil {
```

Comment:
It's unlikely to happen but I think we should protect against PeerCertificates being empty.